iceoryx2-bb/container/src/slotmap.rs

```diff
@@ -0,0 +1,639 @@
+                self.idx_to_data_free_list[entry.previous].next = entry.next;
+            }
+            if entry.next != INVALID {
+                self.idx_to_data_free_list[entry.next].previous = entry.previous;
+            }
+            self.idx_to_data_free_list[idx].next = INVALID;
+            self.idx_to_data_free_list[idx].previous = INVALID;
+        }
+
+        unsafe fn release_free_index(&mut self, idx: usize) {
+            if self.idx_to_data_free_list_head != INVALID {
+                self.idx_to_data_free_list[self.idx_to_data_free_list_head].previous = idx;
+            }
+
+            self.idx_to_data_free_list[idx] = FreeListEntry {
+                previous: INVALID,
+                next: self.idx_to_data_free_list_head,
+            };
+
+            self.idx_to_data_free_list_head = idx;
+        }
+
+        pub(crate) unsafe fn insert_impl(&mut self, value: T) -> Option<SlotMapKey> {
+            match self.acquire_next_free_index() {
+                None => None,
+                Some(key) => {
+                    let key = SlotMapKey(key);
+                    self.store_value(key, value);
+                    Some(key)
+                }
+            }
+        }
+
+        pub(crate) unsafe fn insert_at_impl(&mut self, key: SlotMapKey, value: T) -> bool {
+            self.reserve_index(key.value());
+            self.store_value(key, value)
+        }
+
+        pub(crate) unsafe fn store_value(&mut self, key: SlotMapKey, value: T) -> bool {
+            if key.0 > self.capacity_impl() {
+                return false;
+            }
+
+            let data_idx = self.idx_to_data[key.0];
+            if data_idx != INVALID {
+                self.data[data_idx] = Some(value);
+            } else {
+                let n = self.data_next_free_index.pop_impl().expect("data and idx_to_data correspond and there must be always a free index available.");
+                self.idx_to_data[key.0] = n;
+                self.data[n] = Some(value);
+                self.len += 1;
+            }
+
+            true
+        }
+
+        pub(crate) unsafe fn remove_impl(&mut self, key: SlotMapKey) -> bool {
+            if key.0 > self.idx_to_data.len() {
+                return false;
+            }
+
+            let data_idx = self.idx_to_data[key.0];
+            if data_idx != INVALID {
+                self.data[data_idx].take();
+                debug_assert!(self.data_next_free_index.push_impl(data_idx));
+                self.release_free_index(key.0);
+                self.idx_to_data[key.0] = INVALID;
+                self.len -= 1;
+                true
+            } else {
+                false
+            }
+        }
+
+        pub(crate) fn len_impl(&self) -> usize {
+            self.len
+        }
+
+        pub(crate) fn capacity_impl(&self) -> usize {
+            self.idx_to_data.capacity()
+        }
+
+        pub(crate) fn is_empty_impl(&self) -> bool {
+            self.len_impl() == 0
```

Comment:
Out of curiosity, since you used just `self.len` further above. Is there a reason to use the method in this case?